### PR TITLE
Slight Modification and More Functionalities 

### PR DESCRIPTION
**Pull request by ThisIsXinyuHuo**
1. store the campaigns in the array instead of map
2. create a map: contributor => campaigns
3. modify the struct campaign,  createCampaign() , closeCampaign().
4. add functions: payTo(), withdraw(), getContributors()
5. More functions to be implemented: refund(), requestWithdraw(), requestRefund()
6. getter functions for campaigns need modification.

### CODE REVIEW

File: contracts/Crowdfunding.sol [hunk 1]
```diff
@@ -5,99 +5,220 @@ pragma solidity ^0.8.0;
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 
 contract Crowdfunding {
+    // state variables
+    address private owner;
```
Comment (tonamuhang) [inline contracts/Crowdfunding.sol:9]: Why do we need this?

File: contracts/Crowdfunding.sol [hunk 2]
```diff
@@ -5,99 +5,220 @@ pragma solidity ^0.8.0;
...
-        for (uint256 i = 0; i < numCampaigns; i++) {
-            allCampaigns[i] = campaigns[i];
+        // Check if the goal is met
+        if (
+            campaigns[_campaignId].raisedAmount >= campaigns[_campaignId].goal
+        ) {
+            // Release funds to campaign creator
+            withdraw(_campaignId);
+        } else {
+            refund(_campaignId);
```
Comment (tonamuhang) [inline contracts/Crowdfunding.sol:141]: really makes more sense to be campaign.refund(), but this is how our structure is lol

File: contracts/Crowdfunding.sol [hunk 3]
```diff
@@ -5,99 +5,220 @@ pragma solidity ^0.8.0;
...
+        // an external function call by users to request withdraw
+        // require: state is not CANCEL/CLOSE/SUCCESS; the deadline is passed and raised enough money; the msg.sender is project creator
+        // the campaign will be marked as SUCCESS after requestWithdraw
     }
 
-    // Function to create a new crowdfunding campaign
-    function createCampaign(uint256 _goal, uint256 _deadline) external {
-        require(_goal > 0, "Goal must be greater than 0");
-        require(_deadline > block.timestamp, "Deadline must be in the future");
+    function requestRefund(uint256 _campaignId) external {
```
Comment (tonamuhang) [inline contracts/Crowdfunding.sol:171]: I'm thinking, should we add a modifier that checks if the deadline has passed, and automatically sets the campaign status to closed and refund to contributors if goal is not met?

**Review by tonamuhang — APPROVED** [review]
Thanks Xinyu! LTGM overall, I left some comments for discussion but you can merge this now

File: contracts/Crowdfunding.sol (hunk 2 above)
Comment (ThisIsXinyuHuo) [reply contracts/Crowdfunding.sol:141]: yes. we should create another contract for campaign lol

File: contracts/Crowdfunding.sol (hunk 3 above)
Comment (ThisIsXinyuHuo) [reply contracts/Crowdfunding.sol:171]: yea modifier is a good idea

File: contracts/Crowdfunding.sol (hunk 1 above)
Comment (ThisIsXinyuHuo) [reply contracts/Crowdfunding.sol:9]: forget to delete this line lol

File: contracts/Crowdfunding.sol (hunk 2 above)
Comment (ThisIsXinyuHuo) [reply contracts/Crowdfunding.sol:141]: yes. we should create another contract for campaign lol

**Pull request merged**